### ISSUE TITLE
Option to include key in SourceRecord

### ISSUE THREAD

**Issue by zcox** [confluentinc/kafka-connect-jdbc#28]
Currently it appears that when a new `SourceRecord` is created, the `keySchema` and `key` fields are left null, which results in the message sent to the Kafka topic also having a null key. It would be useful to have the option for the JDBC Connector to use something as the message key, e.g. the primary key id value. If compaction is enabled on this Kafka topic, then the most recent changes to each database table row are replicated in the topic and consumer may then fully rebuild the current state of the table. Otherwise, the message will be removed from the Kafka topic after the retention period passes, and the database change will be lost from the topic.

To achieve this currently, a consumer would need to extract the primary key id out of the message value, and write the key & value to a 2nd Kafka topic.

I'm sure there is complexity around which columns to use as the message key and how to configure this, but it would be nice functionality to have.

**Comment by ewencp** [comment]
@zcox Agreed, this is definitely functionality we want to support! When loading from tables, I think getting the primary key shouldn't be a problem since that information should be available via JDBC. Once you have the right set of columns, converting the values is exactly the same process as it would be for converting a row. One special case we might want to handle is if the primary key is a single column then we can pull it out of the struct and return the column type directly (e.g. if it is an int, don't wrap it in a single-field struct, just return the int with Schema.INT64_SCHEMA directly).

**Comment by zcox** [comment]
@ewencp that all sounds great! :+1:

**Comment by wushujames** [comment]
If you just used a Schema.INT64_SCHEMA directly, then you wouldn't have a column name for the primary key, right?

It just occurred to me that kafka-connect-jdbc has no way to capture deleted rows... You'd have to do soft-deletes (aka add an isDeleted=true column to your database).

**Comment by zcox** [comment]
@wushujames that also occurred to me yesterday, that the jdbc connector cannot detect deleted rows. The only workaround I came up with was what you suggested: using a boolean column to mark a row deleted.

**Comment by malonej7** [comment]
:+1:  Any chance this is going to happen soon?

**Comment by gwenshap** [comment]
It will happen the moment anyone submits a patch :)

**Comment by davidchen322** [comment]
@gwenshap What branch should we work off of if we plan on submitting a patch?  We are currently experimenting with Confluent 2.0.1.

**Comment by nbuesing** [comment]
I am working on a PR for this enhancement, I have enough code in place to where I have POC working. With this being my first contribution to this connector (and confluentinc code in general) I have a few questions.

1)

this change requires GenericDatabaseDialect.describeColumn(rsMetadata, column) to figure out what field is part of the PK, currently, it always returns false. getting PK information is not something you can do from the result set's metadata, but fortunately, you can from the database metadata, and the method `primaryKeyColumns` already exists to do this. 

My question, however, is I do not have the ability to test all dialects to ensure there is not a dialect that doesn't support obtaining this metadata. What is the advice with this?  Do I just state what dialects I was able to test directly and which were only tested via unit tests?

2) 

There are some design decisions I think are best left to those that know the general design of this software more than me. For example, what should the default name for the structure be for the Key (it is table name for the value, and then you can write an SMT to override it).  

Do I:
- keep it as the table name and assume and SMT will be done for key or value 
- add _KEY to the end as in "{table}_KEY" 
- add key to the end such as "{table}Key"?
- I could add another configuration "keySuffix", but I prefer to not do things that could be handled by SMTs, but at the same time, creating key and value with the same struct name seems wrong -- since using an SMT should not be required. I am leaning toward this with _KEY as the default, but someone could change it to "Key".

So, any thoughts or guidelines here would also be helpful.

3. I currently have 3 modes for 'pk.mode' - hoping this behavior aligns with what others have thought for this behavior.

- 'value' -- existing and default behavior where primary key data is ignored and all fields get put in the value's struct (key remains null and it is left to SMTs to extract fields -- again existing behavior)
- 'key' -- the primary key columns go into the struct of the key, and non-primary key fields go into the value's struct
- 'key+value' -- primary key columns go into the key and the value continues to work as-is (has all fields).

**Comment by ewencp** [comment]
1. Been awhile since I've worked on this repo, but there are integration tests, e.g. in https://github.com/confluentinc/kafka-connect-jdbc/tree/master/src/test/java/io/confluent/connect/jdbc/source/integration that should cover at least the most common dialects. Pretty sure they are executed as part of CI, so you should get feedback just by submitting a PR.

For PKs, I'm a bit surprised, I could have sworn that used to work at least for some dbs. Which are you looking at specifically? Derby is used embedded and I thought provided valid info, but we also originally did tests against mysql (admittedly written ~6yrs ago, so I may be misremembering).

2. More configs are usually a pain for users, so it's best to try to find a reasonable default. I'd leave it as the table name and let people work from there. My guess would be that in the vast majority of cases, that will be a reasonable default, and if not, up to them to use an SMT to modify it. There are definitely things that could be improved wrt ergonomics of SMTs, but imo it's better to preserve the raw data and let people munge it (via SMTs, kstreams, or ksqldb) as needed.

3. I'd look for a major version update to potentially make the default `key+value`. Technically it changes behavior, but almost certainly only in a desired way. The only reason I can think of not wanting it is if a table has too many updates and therefore the `key+value` setting routes too much data to one partition (since they would now all hash the same instead of just being round robin on the topic). Is there a specific use case for the `key` option? Obviously it duplicates some data, but since PKs are frequently going to be things like ID columns, a lot of apps would find it easier to just read off the columns from the value rather than having to manually combine key & value, especially in cases where they don't require the PK.
